crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -0,0 +1,600 @@
+        Watch::Directory => Ok((candidate_watch_dir.to_path_buf(), Watch::Directory)),
+    }
+}
+
+/// to allow notify to watch for multiple events (CLOSE_WRITE, CREATE, MODIFY, etc...)
+/// our internal enum `Masks` needs to be converted into a single `WatchMask` via bitwise OR
+/// operations. (Note, our `Masks` type is an enum, `WatchMask` is a bitflag)
+pub(crate) fn pipe_masks_into_watch_mask(
+    masks: &[FileEvent],
+) -> Result<WatchMask, NotifyStreamError> {
+    let mut watch_mask = WatchMask::empty();
+    for mask in masks {
+        watch_mask |= mask.clone().try_into()?
+    }
+    Ok(watch_mask)
+}
+
+impl NotifyStream {
+    /// add a watcher to a file or to a directory
+    ///
+    /// this is implemeted as a direcotry watcher regardless if a file is desired
+    /// to be watched or if a directory. There is an internal data structure that
+    /// keeps track of what is being watched - `self.watchers`
+    /// The `stream` method determines whether the incoming event matches what is
+    /// expected in `self.watchers`.
+    ///
+    /// # Watching directories
+    ///
+    /// ```rust
+    /// use tedge_utils::fs_notify::{NotifyStream, FileEvent, Watch};
+    /// use try_traits::default::TryDefault;
+    /// use std::path::Path;
+    ///
+    /// let dir_path_a = Path::new("/tmp");
+    /// let dir_path_b = Path::new("/etc/tedge/c8y");
+    ///
+    /// let mut fs_notification_stream = NotifyStream::try_default().unwrap();
+    /// fs_notification_stream.add_watcher(dir_path_a, Watch::Directory, &[FileEvent::Created]).unwrap();
+    /// fs_notification_stream.add_watcher(dir_path_b, Watch::Directory, &[FileEvent::Created, FileEvent::Deleted]).unwrap();
+    /// ```
+    ///
+    /// # Watching files
+    ///
+    /// ```rust
+    /// use tedge_utils::fs_notify::{NotifyStream, FileEvent, Watch};
+    /// use tedge_test_utils::fs::TempTedgeDir;
+    /// use try_traits::default::TryDefault;
+    ///
+    /// let ttd = TempTedgeDir::new();  // created a new tmp directory
+    /// let file_a = ttd.file("file_a");
+    /// let file_b = ttd.file("file_b");
+    ///
+    /// let mut fs_notification_stream = NotifyStream::try_default().unwrap();
+    /// fs_notification_stream.add_watcher(ttd.path(), Watch::File(String::from("file_a")), &[FileEvent::Modified]).unwrap();
+    /// fs_notification_stream.add_watcher(ttd.path(), Watch::File(String::from("file_b")), &[FileEvent::Created, FileEvent::Deleted]).unwrap();
+    /// ```
+    /// NOTE:
+    /// in this last example, the root directory is the same: `ttd.path()`
+    /// but the files watched and masks are different. In the background,
+    /// the `add_watcher` fn will add a watch on `ttd.path()` with masks:
+    /// Created, Modified and Deleted. and will update `self.watchers`
+    /// with two entries, one for file_a and one for file_b.
+    ///
+    /// The `stream` method will check that events coming from
+    /// `ttd.path()` match `self.watchers`
+    pub fn add_watcher(
```

Comment:
It's very minor. Just my idea.

This is the current API.
```rust
fs_notification_stream.add_watcher(dir_path_a, Watch::Directory, &[FileEvent::Created]).unwrap(); // for dir
fs_notification_stream.add_watcher(ttd.path(), Watch::File(String::from("file_a")), &[FileEvent::Modified]).unwrap(); // for file
```

I would add
```rust
fs_notification_stream.add_watcher_dir(dir_path_a, &[FileEvent::Created]).unwrap(); // for dir
fs_notification_stream.add_watcher_file(ttd.path(), "file_a", &[FileEvent::Modified]).unwrap(); // for file
```

So, `add_watcher_dir` calls `add_watcher` with `Watch::Directory`.
```rust
pub fn add_watcher_dir(&mut self, dir_path: &Path, masks: &[FileEvent]) -> Result<(), NotifyStreamError>  {
    add_watcher(dir_path, Watch::Directory, masks)
}
```

And `add_watcher_file` calls `add_watcher` with `Watch::File(_)`.
```rust
pub fn add_watcher_file(&mut self, dir_path: &Path, file: &str, masks: &[FileEvent]) -> Result<(), NotifyStreamError>  {
    add_watcher(dir_path, Watch::File(file.to_string()), masks)
}
```

Reply:
This is a nice to have but given the time constraint and that this does not add much value to the user (not the developer), it will be not be addressed in this PR.